### ISSUE TITLE
You must `download()` an article before calling `parse()` on it!

### ISSUE THREAD

**Issue by hoommaah** [codelucas/newspaper#357]
i have a problem with parsing articles and i think its because i placed parse right after downloading the article. do you think there is a chance that the article is not yet done downloading when i started parsing it? any suggestions? thanks!

**Comment by yprez** [comment]
Newspaper is downloading articles synchronously, so that's probably not the issue.
Can you post the code that you're running?

**Comment by yprez** [comment]
Newspaper is downloading articles synchronously, so that's probably not the issue.
Can you post the code that you're running?

**Comment by craymichael** [comment]
I can reproduce this error consistently; it seems as if an Exception should be raised, or maybe even `None` returned instead (i.e. page not recognized as an article or valid page)?

Minimum steps to reproduce (Python 3):
```>>> from newspaper import Article
>>> a = Article('')
>>> a.download()
>>> a.parse()
You must `download()` an article before calling `parse()` on it!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/site-packages/newspaper/article.py", line 168, in parse
    raise ArticleException()
newspaper.article.ArticleException

>>> a = Article('http://www.irishtimes.com/business/sectors/financial-services/ecb-to-reveal-bad-loan-hurdles-for-euro-zone-bank-test-1.1719128')
>>> a.download()
>>> a.parse()
You must `download()` an article before calling `parse()` on it!
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/site-packages/newspaper/article.py", line 168, in parse
    raise ArticleException()
newspaper.article.ArticleException
```

**Comment by toshiro92** [comment]
Same issue for me, I am on Python3.5, and even I called `download()` before `parse()`, it shows me the same error.

**Comment by Helw150** [comment]
@craymichael The Irish Times link you placed there is 404ing, which will intentionally cause this error.

However, I do have several instances of non-404ing links that raise this error. I believe it has something to do with pages with slow load times.

**Comment by codelucas** [comment]
I'll take a look later, thanks for reporting this guys

**Comment by codelucas** [comment]
This is being addressed right now in https://github.com/codelucas/newspaper/pull/378

**Comment by codelucas** [comment]
Hey all, happy to announce that this has been FIXED in the latest version of newspaper3k in pip (version 2.0.1) pull now!

```
FAIL EXAMPLE: 

>>> a = Article('http://www.irishtimes.com/business/sectors/financial-services/ecb-to-reveal-bad-loan-hurdles-for-euro-zone-bank-test-1.1719128')
>>> a.download()
>>> a.parse()
Article `download()` on URL http://www.irishtimes.com/business/sectors/financial-services/ecb-to-reveal-bad-loan-hurdles-for-euro-zone-bank-test-1.1719128 failed with 404 Client Error: Not Found for url: http://www.irishtimes.com/business/sectors/financial-services/ecb-to-reveal-bad-loan-hurdles-for-euro-zone-bank-test-1.1719128
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lucasou-yang/Documents/github_experiments/test-newspaper-env/lib/python3.6/site-packages/newspaper/article.py", line 185, in parse
    self.throw_if_not_downloaded_verbose()
  File "/Users/lucasou-yang/Documents/github_experiments/test-newspaper-env/lib/python3.6/site-packages/newspaper/article.py", line 525, in throw_if_not_downloaded_verbose
    raise ArticleException()
newspaper.article.ArticleException
>>>
>>>

WORKING EXAMPLE:

>>> a = Article('http://www.cnn.com/2017/06/10/politics/donald-trump-obama-presidents-club/index.html')
>>> a.download()
>>> a.parse()
>>>
>>> a.title
"Trump's strain with Obama marks departure from presidential fraternity"
>>> a.html[:100]
'<!DOCTYPE html><html class="no-js"><head><meta content="IE=edge,chrome=1" http-equiv="X-UA-Compatibl'
>>> a.text[:100]                                                                                                                                                                     
'Washington (CNN) Nearly five months ago, President Donald Trump bid farewell to a grinning Barack Ob'
>>> a.movies
[]
>>> a.authors
['Kevin Liptak', 'Cnn White House Producer']
>>> a.top_image
'http://i2.cdn.cnn.com/cnnnext/dam/assets/170508112504-03-obama-trump-oval-office-1110-super-tease.jpg'
```

**Comment by hoommaah** [comment]
WOW! Thanks, mate!

**Comment by mahdi-saberi** [comment]
it still has errors
e.g.
http://khabareazad.ir/fa/news/2561/%D9%88%D8%B1%D8%B2%D8%B4%DA%AF%D8%A7%D9%87%20%D9%85%D9%84%DB%8C%20%D9%BE%DA%A9%D9%86/

**Comment by dineshdaultani** [comment]
Hi all,
I was still getting this error again and again. I don't know why this issue is there, but I think it's something related to the time taken between parse() function call and article usage statement.
For ex:
`article = Article(url)`
`article.download()`
`article.parse()`
`article.text() # Error generated here `

I have manually found a workaround for this error. I am no longer getting the same error by doing the workaround as below:

`article = Article(url)`
`article.download()`
`article.parse()`
`time.sleep(1) # "import time"  `
`article.text() # Error generated here` 

Basically, I am pausing the Python Kernel for a second to download and parse the article properly.

Hope this helps. Let me know if this fixes your problem as well.

**Comment by jessecooper** [comment]
I ended up having to put as much as a 2 second wait for the parse to function properly. I ended up using the following code:
```
...
article = Article(url)
article.download()
while article.download_state == 0: #ArticleDownloadState.NOT_STARTED is 0
    time.sleep(1)
article.parse()
...
```

**Comment by pouta** [comment]
@jessecooper After using your workaround I still got some errors. After digging into the source code I found the article has a success flag so it made more sense to wait for the article to be successfully downloaded.

I hope people find this snipet useful!

```
article = Article(url)
article.download()
while article.download_state != 2: #ArticleDownloadState.SUCCESS is 0
    time.sleep(1)
    article.parse()
```

**Comment by The-Gupta** [comment]
Hey @pouta @jessecooper @codelucas, 
With this link - https://www.expertily.com/blog/narayan-murthy-slams-infosys
I get the error in bold. Even though it got downloaded, article.download_state is 0.
I updated the library.

from newspaper import Article
import time

url='https://www.expertily.com/blog/narayan-murthy-slams-infosys'

article = Article(url)
article.download()

article.download_state
Out[4]: 0

article.parse()
**You must `download()` an article first!
Traceback (most recent call last):

  File "<ipython-input-5-ad4c0cca82d1>", line 1, in <module>
    article.parse()

  File "D:\Anaconda3\envs\theResearcher\lib\site-packages\newspaper\article.py", line 185, in parse
    self.throw_if_not_downloaded_verbose()

  File "D:\Anaconda3\envs\theResearcher\lib\site-packages\newspaper\article.py", line 522, in throw_if_not_downloaded_verbose
    raise ArticleException()

ArticleException**

**Comment by xzmxzm007** [comment]
My url is a url list consist of several urls, I use the following code to solve the problem.
I hope my code can help you!
```
# save result in dic
news={}
n=0
# the process will show the useless url without stopping the algorithm and show the pencentile
while n < len(url):
    article=Article(url[n])
    article.download()
    while article.download_state != 2:   
        print(n)
        time.sleep(1)
        break
    if article.download_state == 2:
        article.parse()
        article.nlp()
        news[n]=article.summary
        keywords[n]=article.keywords
        title[n]=article.title
    n+= 1
    for i in range(100):
        k = i + 1
        str = '>' * (i // 2) + ' ' * ((100 - k) // 2)
        sys.stdout.write('\r' + str + '[%s%%]' % (i + 1))
        sys.stdout.flush()
        time.sleep(0.1)
```
